Apps/FlightSoftware/Watchdog2/src/main.c

```diff
@@ -0,0 +1,241 @@
+#include <msp430.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "include/buffer.h"
+#include "include/uart.h"
+#include "include/bsp.h"
+#include "include/adc.h"
+#include "include/flags.h"
+#include "include/i2c.h"
+#include "include/ip_udp.h"
+#include "include/watchdog.h"
+
+
+/* define all of the buffers used in other files */
+__volatile struct buffer uart0rx, uart0tx, uart1rx, uart1tx, hercbuf;
+__volatile uint16_t loop_flags;
+extern uint8_t heating;
+uint8_t lastHeater = 0;
+uint8_t heatingControlEnabled = 1;
+
+// @suppress("Invalid arguments")
+
+/* function definitions in ground_cmd.c */
+void parse_ground_cmd(struct buffer *pp);
+void send_earth_heartbeat();
+
+/**
+ * main.c
+ */
+#pragma PERSISTENT(rovstate)
+enum rover_state rovstate = RS_KEEPALIVE;
+
+void uart1_disable();
+void uart0_init();
+
+void enterMode(enum rover_state newstate) {
+    switch (newstate) {
+    case RS_SLEEP:
+        /* fallthrough to keepalive */
+    case RS_SERVICE:
+        /* service is same as keep alive */
+        /* service mode is a temporary step transition to mission mode */
+    case RS_KEEPALIVE:
+        /* power everything off and set resets */
+        setRadioReset();
+        setFPGAReset();
+        setMotorsReset();
+        setHerculesReset();
+        powerOffFpga();
+        powerOffMotors();
+        powerOffRadio();
+        powerOffHercules();
+//        fuelGaugeLowPower();
+        /* TODO: do we want to do it in this order? */
+
+        /* turn off voltage rails */
+        disable3V3PowerRail();
+        disable24VPowerRail();
+//        disableBatteries();
+        /* monitor only lander voltages */
+        adc_setup_lander();
+        enableBatteries();      // need enable batteries to read from fuel gauge
+        enableHeater();
+        startChargingBatteries();
+        break;
+    default:
+    case RS_MISSION:
+        /* bootup process - enable all rails */
+        enable3V3PowerRail();
+        enable24VPowerRail();
+        enableBatteries();
+
+        /* swap uart's */
+//        uart1_disable(); // enabled: 0.0467
+        uart0_init();
+
+        /* start monitoring only mission-relevant voltages */
+        adc_setup_mission();
+
+        /* power everything on and release resets */
+        powerOnHercules();
+        releaseHerculesReset();
+        powerOnFpga();
+        powerOnMotors();
+        powerOnRadio();
+        releaseRadioReset();
+        releaseFPGAReset();
+        stopChargingBatteries();
+
+        __delay_cycles(1234567); //give fuel gauge time to start up
```

Comment:
???

Reply:
In a newer version of the watchdog code, this line is now:
`__delay_cycles(12345678); //give fuel gauge [50 ms] & wifi [~750 ms] time to start up`
although there's a magic number and the delays have not be characterized particularly well. Would a \#define WIFI_STARTUP_DELAY in main.h be acceptable?

Reply:
Do we have any way other than time to check if these things have started up? Ideally we wouldn't use time at all. The new comment definitely helps though

Reply:
That's a really good question. @justinnuwin how are you checking for when the wifi is ready in the hercules?